openedx/adg/lms/applications/tests/test_analytics.py

```diff
@@ -0,0 +1,277 @@
+    else:
+        program_prereq_emit.assert_not_called()
+
+
+def add_course_grade_for_user(user, course, is_passed=False):
+    """
+    Create a course enrollment and compute course grade for the given course
+    and user pair. Grade is passed or failed based on the `is_passed` value
+
+    Arguments:
+        user (User): User to compute the grade for
+        course (CourseOverview): Course to compute the grade for
+        is_passed (Boolean): Should the user pass in the course or not
+
+    Returns:
+        PersistentCourseGrade: Computed grade based on the given information
+    """
+    CourseEnrollmentFactory(user=user, course=course)
+    kwargs = {
+        'passed': is_passed,
+        'percent_grade': 100 if is_passed else 0,
+        'letter_grade': 'Pass' if is_passed else ''
+    }
+    grade_achieved = PersistentCourseGrade.update_or_create(user.id, course.id, **kwargs)
+    return grade_achieved
+
+
+@pytest.mark.parametrize('is_passed', [True, False])
+@pytest.mark.django_db
+def test_course_completion_emit_when_user_passes_or_fails(is_passed, user, courses, mocker):
+    """
+    Test that the course completion event is emitted when a user completed a course and
+    grade is computed
+    """
+    mock_testing_env = mocker.patch('lms.djangoapps.grades.models.is_testing_environment')
+    mock_testing_env.return_value = False
+    mock_course_completed_emit = mocker.patch('lms.djangoapps.grades.models.emit_course_progress_event')
+
+    course_to_grade = courses['test_course1']
+    grade_achieved = add_course_grade_for_user(user, course_to_grade, is_passed=is_passed)
+    mock_course_completed_emit.assert_called_once_with(user, course_to_grade, grade=grade_achieved)
+
+
+@pytest.mark.parametrize('is_passed, is_program_prereq', [(True, True), (True, False), (False, True), (False, False)])
+@pytest.mark.django_db
+def test_program_prereq_course_progress_emit_for_completion(is_passed, is_program_prereq, user, mocker):
+    """
+    Test that program prereq completed event is emitted when a program prereq course is completed and
+    passed successfully
+    """
+    mock_course_language = mocker.patch('openedx.adg.lms.applications.analytics.get_language_info')
+    mock_course_language.return_value = {'name': 'English'}
+    mock_all_modules_attempted = mocker.patch('openedx.adg.lms.applications.helpers.has_attempted_all_modules')
+    mock_all_modules_attempted.return_value = True
+    mock_program_prereq_emit = mocker.patch(
+        'openedx.adg.lms.applications.analytics.emit_enrollment_event_for_program_prereq_courses'
+    )
+
+    course_group = MultilingualCourseGroupFactory(is_program_prerequisite=is_program_prereq)
+    course_to_grade = MultilingualCourseFactory(multilingual_course_group=course_group).course
+    grade_achieved = add_course_grade_for_user(user, course_to_grade, is_passed=is_passed)
+
+    emit_course_progress_event(user, course_to_grade, grade=grade_achieved)
```

Comment:
Why are we explicitly emitting an event here? Wouldn't this function be automatically called, by signal handler, upon completion?

Reply:
The signal handler triggers this function on every enrollment and grade computation. And this function then decides to emit an event if the course in a program prereq course.

The former logic is already tested and the latter is being tested here.